### ISSUE TITLE
Kokkos_Random: Random number sequence seems to get repeated for different threads on CUDA

### ISSUE THREAD

**Issue by Shihab-Shahriar** [kokkos/kokkos#6140]
I can't provide a fully reproducible minimal code at this moment, but here's the scenario: I am trying to initialize say a million particles and set their properties by generating e.g. 7 different doubles for each particle. 

```
Kokkos::Random_XorShift64_Pool<Kokkos::CudaSpace> random_pool(12345);

// 2000 teams, 512 threads per team
Kokkos::parallel_for(teamPolicy,
        KOKKOS_LAMBDA(const TeamHandleType& team)
        {
            unsigned numElems = 512;
            Kokkos::parallel_for(Kokkos::TeamThreadRange(team, 0u, numElems), [&] (const int& i)
            {
                Kokkos::Random_XorShift64<stk::ngp::MemSpace> gen = random_pool.get_state();

                const double x1 = gen.drand(0.0, 1.0);
                const double x2 = gen.drand(0.0, 1.0);
                ...
                const double x7 = gen.drand(0.0, 1.0);

                random_pool.free_state(gen);
           }
     }
```

Sometimes, I would get two particles with exactly same values for all randomly generated properties. It doesn't happen very often- once in every 3 or 4 runs of simulation maybe, and I have only seen it when dealing with around 1million or higher number of particles.

Edit: I do have an unique id for each particle that can be used as a seed to generate the sequences independently for each particle. Is there any Kokkos API that allows that? or any plan to do this in future?

**Details**
1. Modules: `GCC/10.2.0 CUDA/11.2.1 OpenMPI/4.0.5 CMake/3.23.1`
2. Kokkos release: 4.0.1
3. platform: CUDA and linux.
4. cmake configure command:
```
cmake ${ROOT_DIR} \
    -DCMAKE_CXX_COMPILER="${Trilinos_Source_DIR}/packages/kokkos/bin/nvcc_wrapper" \
    -DCMAKE_CXX_FLAGS="-O0 -lmpi" \
    -DTrilinos_DIR:PATH=${Trilinos_Install_DIR} \
    -DKokkos_ROOT:PATH="${Trilinos_Install_DIR}" \
    -DCMAKE_BUILD_TYPE:STRING=Debug \
    -DKokkos_ENABLE_CUDA=ON
```

**Comment by Shihab-Shahriar** [comment]
pinging @ajpowelsnl @crtrott

**Comment by crtrott** [comment]
Hm, the stuff is already getting initalized with different numbers, I actually am wondering if we might miss a memory fence for the "free_state" and so a second thread getting the same generator, may still see the start value of the thread who previously grepped that generator.

**Comment by crtrott** [comment]
Yeah its missing memory fences :-(

**Comment by ajpowelsnl** [comment]
@fnrizzi, @crtrott - have the fixes for this issue been merged yet?

**Comment by fnrizzi** [comment]
Not yet

**Comment by ajpowelsnl** [comment]
@Shihab-Shahriar , @fnrizzi -- did #6290 resolve the problem?  If so, can we close the issue?

**Comment by Shihab-Shahriar** [comment]
Yes, I think we can.

**Comment by PhilMiller** [comment]
Wait, no, this hasn't been released yet. The bug existed in a prior release, and a new release with the fix hasn't been shipped.

**Comment by masterleinad** [comment]
Fixed in #6290.
